### PR TITLE
implement on_schema_change

### PR DESCRIPTION
**Pull request by AdamE87**
## Overview
Implementation of on_schema_change for incremental models

- Description:
- Related issue(s): #48
-
## Checklist
- [ ] `CHANGELOG.md` updated and added information about my change

### CODE REVIEW

**Comment by AdamE87** [comment]
My first time working with tox, so I ran the tests and the unit tests passed but the integration tests failed and i'm not totally sure on how to resolve. Help would be appreciated.

**Comment by hovaesco** [comment]
That's an error you are getting:
```
08:45:38  Runtime Error in model dim_order_dates_custom_schema (models/marts/core/dim_order_dates_custom_schema.sql)
08:45:38    TrinoUserError(type=USER_ERROR, name=TABLE_NOT_FOUND, message="line 2:30: Table 'memory.default.dim_order_dates_custom_schema__dbt_tmp' does not exist", query_id=20220414_084538_00215_pyttn)
```

**Comment by AdamE87** [comment]
Every time I run the Makefile it's complaining about not being able to find a particular network, which i'm guessing is dbt-net. Thing is the make file creates the network, but the terminal it's looking for a very specific network id, the same one each time. Any ideas?

**Comment by hovaesco** [comment]
try to create a network using `docker network create dbt-net`, perhaps clean networks before using `docker network prune -f`

**Comment by AdamE87** [comment]
Yeah tried both and no luck

**Comment by AdamE87** [comment]
Modified the `init_trino.bash` file to --force-recreate on the `docker-compose-trino.yml` up (line 11). Resolved the issue

**Comment by AdamE87** [comment]
Right i'm trying to figure out what's happened here. The error message states that it's trying to look for `memory.default.dim_order_dates_custom_chema__dbt_tmp` but that model if actually on a custom schema so should be looking for `memory.default_custom_schema.dim_....`. I'm not sure if the changes i've proposed are causing this error or not!

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 1]
```diff
@@ -1,17 +1,71 @@
...
+
+    {%- set dest_cols_csv = get_quoted_csv(dest_columns | map(attribute='name')) -%}
 
-    {%- set dest_columns = adapter.get_columns_in_relation(target_relation) -%}
-    {%- set dest_cols_csv = dest_columns | map(attribute='quoted') | join(', ') -%}
     insert into {{ target_relation }}
-    select {{dest_cols_csv}} from {{ tmp_relation.include(database=true, schema=true) }};
+    select {{dest_cols_csv}} from {{ source_relation.include(database=false, schema=false) }}
 
-    drop table if exists {{ tmp_relation }};
```
Comment (hovaesco) [inline dbt/include/trino/macros/materializations/incremental.sql:8]: Perhaps that's the culprit of test failures.

Comment (AdamE87) [reply dbt/include/trino/macros/materializations/incremental.sql:8]: Right, these changes come from @findinpath in his feature/incremental branch which this was forked from. 

Should I re-do the changes forking from master instead?

Comment (hovaesco) [reply dbt/include/trino/macros/materializations/incremental.sql:8]: Try to add this line to your code and see if that helps.

Comment (AdamE87) [reply dbt/include/trino/macros/materializations/incremental.sql:8]: Yep, that did the trick. Thanks for your help @hovaesco

File: dbt/include/trino/macros/materializations/incremental.sql [hunk 2]
```diff
@@ -1,17 +1,73 @@
...
+
+
+{% macro dbt_trino_get_insert_into_sql(source_relation, target_relation, dest_columns) %}
+
+    {%- set dest_cols_csv = get_quoted_csv(dest_columns | map(attribute='name')) -%}
 
-    {%- set dest_columns = adapter.get_columns_in_relation(target_relation) -%}
-    {%- set dest_cols_csv = dest_columns | map(attribute='quoted') | join(', ') -%}
     insert into {{ target_relation }}
-    select {{dest_cols_csv}} from {{ tmp_relation.include(database=true, schema=true) }};
```
Comment (hovaesco) [inline dbt/include/trino/macros/materializations/incremental.sql:6]: Why have you changed `tmp_relation.include(database=true, schema=true)` to `source_relation.include(database=false, schema=false)` ? 

I believe `(database=false, schema=false)` causes tests to fail.

**Comment by hovaesco** [comment]
@mdesmet some of your changes from https://github.com/starburstdata/dbt-trino/pull/62 are here. Could you please take a look on it?

File: dbt/include/trino/macros/materializations/incremental.sql (hunk 2 above)
Comment (AdamE87) [reply dbt/include/trino/macros/materializations/incremental.sql:6]: I haven't, these were changes from the branch I forked `feature/incremental` .

Comment (hovaesco) [reply dbt/include/trino/macros/materializations/incremental.sql:6]: Probably you're just missing changes from https://github.com/starburstdata/dbt-trino/pull/39. 
Try to change `(database=false, schema=false)` to `(database=true, schema=true)`

Comment (AdamE87) [reply dbt/include/trino/macros/materializations/incremental.sql:6]: Yeah that works great

**Comment by mdesmet** [comment]
> @mdesmet some of your changes from #62 are here. Could you please take a look on it?

I think @AdamE87 mentioned he took changes from feature/incremental, which is some backup branch related to #1. So we should probably remove the changes that are related to adding the "insert_overwrite" strategy and focus only on adding the `on_schema_change` feature here.

IMHO it also necessary to have some integration tests that cover the `on_schema_change` feature. See [dbt docs](https://docs.getdbt.com/docs/building-a-dbt-project/building-models/configuring-incremental-models#what-if-the-columns-of-my-incremental-model-change). Especially the `append_new_columns` and `sync_all_columns` should be tested.

**Comment by AdamE87** [comment]
@mdesmet totally agree on the testing front, creating the tests is not something i'm experienced in by of course happy to learn. I do have some concerns about the `on_schema_change` feature for some of the connectors though. MinIO for example does not allow add or drop columns, it's a total table drop and recreate.

**Comment by mdesmet** [comment]
> @mdesmet totally agree on the testing front, creating the tests is not something i'm experienced in by of course happy to learn. I do have some concerns about the `on_schema_change` feature for some of the connectors though. MinIO for example does not allow add or drop columns, it's a total table drop and recreate.

Don't think it's necessarely linked to MinIO but rather wich connector you use. Currently we only have the `memory` connector in dbt-trino. This conector does not support `ALTER TABLE ADD/DROP column`. So test cases might be difficult for now.

**Comment by AdamE87** [comment]
Hive connector is the one I use for Minio, but I take your point.

**Comment by hovaesco** [comment]
I have created an issue to migrate all tests from memory connector to PostgreSQL https://github.com/starburstdata/dbt-trino/issues/66 I will just start to do this and once it's done @AdamE87 you could start working on tests. 

Moreover, we are going to start using a new adapter testing framework https://github.com/starburstdata/dbt-trino/pull/65, which is easier and more clear to understand. From now on, we should only create tests based on it.

**Comment by AdamE87** [comment]
Great stuff, i'll begin to familarise myself with tests already written in there.

**Comment by hovaesco** [comment]
@AdamE87 PostgreSQL can be now used in tests https://github.com/starburstdata/dbt-trino/pull/67. So you can tests features like `append_new_columns` and `sync_all_columns`.

**Comment by mdesmet** [comment]
Closed as #136 has been merged.

**Pull request closed without merging**